### PR TITLE
Dag processor manager queue split (fixes SLAs)

### LINKED ISSUE #15596: Using SLAs causes DagFileProcessorManager timeouts and prevents deleted dags from being recreated
**Apache Airflow version**: 2.0.1 and 2.0.2

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): N/A

**Environment**: Celery executors, Redis + Postgres

- **Cloud provider or hardware configuration**: Running inside docker
- **OS** (e.g. from /etc/os-release): Centos (inside Docker)

**What happens**:

In 2.0.0 if you delete a dag from the GUI when the `.py` file is still present, the dag is re-added within a few seconds (albeit with no history etc. etc.). Upon attempting to upgrade to 2.0.1 we found that after deleting a dag it would take tens of minutes to come back (or more!), and its reappearance was seemingly at random (i.e. restarting schedulers / guis did not help).

It did not seem to matter which dag it was.

The problem still exists in 2.0.2.

**What you expected to happen**:

Deleting a dag should result in that dag being re-added in short order if the `.py` file is still present.

**Likely cause**

I've tracked it back to an issue with SLA callbacks. I strongly suspect the fix for Issue #14050 was inadvertently responsible, since that was in the 2.0.1 release. In a nutshell, it appears the dag_processor_manager gets into a state where on every single pass it takes so long to process SLA checks for one of the dag files that the entire processor times out and is killed. As a result, some of the dag files (that are queued behind the poison pill file) never get processed and thus we don't reinstate the deleted dag unless the system gets quiet and the SLA checks clear down.

To reproduce in _my_ setup, I created a clean airflow instance. The only materially important config setting I use is `AIRFLOW__SCHEDULER__PARSING_PROCESSES=1` which helps keep things deterministic.

I then started adding in dag files from the production system until I found a file that caused the problem. Most of our dags do not have SLAs, but this one did. After adding it, I started seeing lines like this in `dag_processor_manager.log` (file names have been changed to keep things simple)

```
[2021-04-29 16:27:19,259] {dag_processing.py:1129} ERROR - Processor for /home/airflow/dags/problematic.py with PID 309 started at 2021-04-29T16:24:19.073027+00:00 has timed out, killing it.
```
Additionally, the stats contained lines like:
```
File Path                                                            PID  Runtime      # DAGs    # Errors  Last Runtime    Last Run
-----------------------------------------------------------------  -----  ---------  --------  ----------  --------------  -------------------
/home/airflow/dags/problematic.py                                    309  167.29s           8           0  158.78s         2021-04-29T16:24:19
```
(i.e. 3 minutes to process a single file!)

Of note, the parse time of the affected file got longer on each pass until the processor was killed. Increasing `AIRFLOW__CORE__DAG_FILE_PROCESSOR_TIMEOUT` to e.g. 300 did nothing to help; it simply bought a few more iterations of the parse loop before it blew up.

Browsing the log file for `scheduler/2021-04-29/problematic.py.log` I could see the following:

<details><summary>Log file entries in 2.0.2</summary> 

```
[2021-04-29 16:06:44,633] {scheduler_job.py:629} INFO - Processing file /home/airflow/dags/problematic.py for tasks to queue
[2021-04-29 16:06:44,634] {logging_mixin.py:104} INFO - [2021-04-29 16:06:44,634] {dagbag.py:451} INFO - Filling up the DagBag from /home/airflow/dags/problematic
[2021-04-29 16:06:45,001] {scheduler_job.py:639} INFO - DAG(s) dict_keys(['PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-S1-weekends', 'PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends', 'PARQUET-BASIC-DATA-PIPELINE-TODAY-S1-weekends', 'PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends', 'PARQUET-BASIC-DATA-PIPELINE-TODAY-S2-weekends', 'PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-S2-weekends', 'PARQUET-BASIC-DATA-PIPELINE-TODAY-S3-weekends', 'PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-S3-weekends']) retrieved from /home/airflow/dags/problematic.py
[2021-04-29 16:06:45,001] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends
[2021-04-29 16:06:46,398] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends
[2021-04-29 16:06:47,615] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends
[2021-04-29 16:06:48,852] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends
[2021-04-29 16:06:49,411] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends
[2021-04-29 16:06:50,156] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends
[2021-04-29 16:06:50,845] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-SP500_Index_1-weekends
[2021-04-29 16:06:52,164] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends
[2021-04-29 16:06:53,474] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-YESTERDAY-APPEND-S1-weekends
[2021-04-29 16:06:54,731] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-SP500_Index_1-weekends
[2021-04-29 16:06:55,345] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends
[2021-04-29 16:06:55,920] {scheduler_job.py:396} INFO - Running SLA Checks for PARQUET-BASIC-DATA-PIPELINE-TODAY-APPEND-S1-weekends

and so on for 100+ more lines like this... 
```

</details>

Two important points: from the above logs:
1. We seem to be running checks on the same dags multiple times
2. The number of checks grows on each pass (i.e. the number of log lines beginning "Running SLA Checks..." increases on each pass until the processor manager is restarted, and then it begins afresh)

**Likely location of the problem**:

This is where I start to run out of steam. I believe the culprit is this line: https://github.com/apache/airflow/blob/2.0.2/airflow/jobs/scheduler_job.py#L1813

It seems to me that the above leads to a feedback where each time you send a dag callback to the processor you include a free SLA callback as well, hence the steadily growing SLA processing log messages / behaviour I observed. As noted above, this method call _was_ in 2.0.0 but until Issue #14050 was fixed, the SLAs were ignored, so the problem only kicked in from 2.0.1 onwards.

Unfortunately, my airflow-fu is not good enough for me to suggest a fix beyond the Gordian solution of removing the line completely (!); in particular, it's not clear to me how / where SLAs _should_ be being checked. Should the dag_processor_manager be doing them? Should it be another component (I mean, naively, I would have thought it should be the workers, so that SLA checks can scale with the rest of your system)? How should the checks be enqueued? I dunno enough to give a good answer. 🤷 

**How to reproduce it**:

In our production system, it would blow up every time, immediately. _Reliably_ reproducing in a clean system depends on how fast your test system is; the trick appears to be getting the scan of the dag file to take long enough that the SLA checks start to snowball. The dag below did it for me; if your machine seems to be staying on top of processing the dags, try increasing the number of tasks in a single dag (or buy a slower computer!)

<details><summary>Simple dag that causes the problem</summary>

```
import datetime as dt
import pendulum

from airflow import DAG
from airflow.operators.bash import BashOperator


def create_graph(dag):
    prev_task = None
    for i in range(10):
        next_task = BashOperator(
            task_id=f'simple_task_{i}',
            bash_command="echo SLA issue",
            dag=dag)

        if prev_task:
            prev_task >> next_task

        prev_task = next_task


def create_dag(name: str) -> DAG:
    tz_to_use = pendulum.timezone('UTC')

    default_args = {
        'owner': 'your.email.here@email.co',
        'start_date': dt.datetime(2018, 11, 13, tzinfo=tz_to_use),
        'email': ['sla_fail@some.domain.or.other'],
        'email_on_failure': False,
        'email_on_retry': False,
        'sla': dt.timedelta(hours=13),
    }

    dag = DAG(name,
              catchup=False,
              default_args=default_args,
              max_active_runs=10,
              schedule_interval="* * * * *")

    create_graph(dag)

    return dag

for i in range(100):
  name = f"sla_dag_{i}"
  globals()[name] = create_dag(name)
```

</details>

To reproduce:
1. Configure an empty airflow instance, s.t. it only has one parsing process (as per config above).
2. Add the file above into the install. The file simply creates 100 near-trivial dags. On my system, airflow can't stay ahead, and is basically permanently busy processing the backlog. Your cpu may have more hamsters, in which case you'll need to up the number of tasks and/or dags.
2. Locate and tail the `scheduler/[date]/sla_example.py.log` file (assuming you called the above `sla_example.py`, of course)
3. This is the non-deterministic part. On my system, within a few minutes, the processor manager is taking noticeably longer to process the file and you should be able to see lots of SLA log messages like my example above ☝️. Like all good exponential growth it takes many iterations to go from 1 second to 1.5 seconds to 2 seconds, but not very long at all to go from 10 seconds to 30 to 💥 

**Anything else we need to know**:

1. I'm working around this for now by simply removing the SLAs from the dag. This solves the problem since the SLA callbacks are then dropped. But SLAs are a great feature, and I'd like them back (please!).
2. Thanks for making airflow and thanks for making it this far down the report!

### PR DESCRIPTION
**Pull request by argibbs**
# The "I'm determined to fix SLAs" PR

OK, so #25147 made a start in this direction. Summing up the, er, summary from _that_ MR, the problem was that SLA callbacks could keep occuring, and prevent the dag processor manager from ever processing more than 2 or 3 dags in the queue before the SLA callbacks re-upped and went to the front of the queue.

Under the new behaviour, the SLA callbacks went to the _back_ of the queue. This guaranteed that the queue would be processed at least once. However, it turns out that dags on disk would only be re-added to the queue _once the queue was empty_. But with SLA callbacks arriving All. The. Time. the queue would never drain, and we'd never re-read dags from disk. So if you updated the dag file, you'd have to bounce the scheduler to pick up the change, and then it would process all non-SLA-generating DAGs exactly once. And then you'd need to bounce again.

### Related Issues

Closes #15596 (I hope!)

# I've almost certainly missed some steps.

Before I go into a bit more detail about the change, I'd like to acknowledge that as a (very) small time contributor to the project, I'm not familiar with all the done things when making more radical changes. In particular, I assume there's more doc changes needed than just a newsfragment. I've added a config flag, some metrics, _and_ the behaviour of the queue processing has subtly changed (for the better I hope!)

I'd very much appreciate someone(s) leaving a comment :point_down: telling me what else I need to do in terms of docs etc.

# TL;DR

I mean, I see your point. Skip to the bottom for the summary.

# Pay attention, here comes the science bit

Ok, so to briefly recap the queue behaviour prior to this change:
1. The scheduler spins up.
2. The worker loop starts.
3. It scans the disk to create a list of all known files. _It does not add these files to the queue_.
4. If the queue is empty (on start up it will be) then all known files are added to the queue.
5. It then spins up dag processors (via multiprocessing or in-line, depending on set up) to consume the first N entries in the queue (where N is a function of config settings, and includes a count of any files currently being processed from a previous pass of the loop)
6. It then goes back to step 2.

- Note that Step 3 is re-run periodically based on a time interval set in config (`dag_dir_list_interval`). It does not require that the queue be empty. Any changes to the set of files on disk e.g. dags being deleted will cause the manager to remove dags from the queue.
- After a few iterations of the loop, the contents of the queue are processed, and step 4 kicks in, and the dag files present on disk are re-added to the queue again.
    - The default config is to do this no more than once every 30 seconds (`min_file_process_interval`), so if your dag files only take 10 seconds to process, there will be 20 seconds of idle time, but if your dag files take a minute to process, then the manager will be permanently busy, because as soon as the queue drains, it'll be well past time to reload the queue from disk.
    - This only works if the queue _actually empties_ - as foreshadowed above, if the queue never empties (because SLAs) then step 4 will never kick in, and we'll never reload the queue. Step 3 will ensure that within a short interval any deleted dags are removed from the queue (should they happen to be one of the ones generating SLAs) but updated and newly added dag files will not be picked up.
    - To be clear, this is the problem I alluded to in my comments at the end of the previous MR after it got merged.

Locally, I tested a hacky fix whereby on receipt of a SLA callback I still add the callback, but I simply _didn't add the dag to the queue_ (it's a one-line change - extremely simple!). This works, but means that SLA callbacks are only processed when the queue drains, and is reprocessed (because then all dags are added to the queue, guaranteeing that we will process any outstanding SLA callbacks). However, If someone has specified a large wait time between loading dags from disk, this will affect how timely the SLA alerts are, which is fine for me, because I don't do that, but I wasn't getting a "this will be fine for everyone" vibe from the change (I did say it was hacky!).

Also, there's another catch. While the problem is much more prevalent with SLAs, these are not the only callbacks. I could envisage a situation where someone configures a dag with a very small interval (e.g. a dag run every 10 seconds). While I think this is a much more theoretical problem that might not ever exist in the wild - this isn't really the use case Airflow is intended for - the upshot would be that a dag generating lots of dag callbacks would be spamming the queue. And those callbacks still go to the _front_ of the queue, i.e. you're back to the situation I tried to solve in the previous MR!

# I don't think that word means what you think it means

I decided that fundamentally, part of the the problem was that the queue should be FIFO. And it wasn't. But if I made it FIFO, then the higher priority DAG callbacks would have to wait their turn behind the dag files loaded from disk, and I'm pretty sure stopping that would eliminate some of the speed-ups Airflow 2 was trumpeted as solving. Airflow 1 used to have on average a 15 second gap (= 30/2) between one task completing and the downstream tasks being scheduled, because once the task completed, you had to wait for the manager to drain the queue, add the files to the queue from disk, and then process the dag. (And that's assuming you could even process all your dags in &lt;30 seconds...). In Airflow 2, because of the dag callbacks, the gap between downstream tasks being scheduled is usually sub-second.

I didn't want to be the guy who accidentally breaks _that_ particular speed up. :scream: 

So I did two things:

## Thing 1: Tackling the FIFO issue aka gazumping callbacks.
1. I split the existing queue into two: a priority queue, and a standard queue. the priority queue is drained before the standard queue.
2. You can probably guess what goes where, but: the dag callbacks go into the priority queue, while the SLAs callbacks and dags found on disk are added to the standard queue.
3. **Both queues are FIFO.** This means that rapidly updating dags can't gazump other slightly-less-rapidly updating dags, because they go to the back. Of course, most of the time, there's only one dag with callbacks in the priority queue, and there's no material change in behaviour. 
4. Hypothetically, the priority queue could be permanently busy. If this happens, it'll be like it was with the SLAs; the standard queue would never get a look-in, and we'd stop processing new/updated dags.
5. So **I also added a new config param**, This is called `max_file_process_interval` and it's the dual to the existing `min_file_process_interval`. It guarantees that if you do happen to have a permanently busy priority queue, eventually we'll take a breather, and process the files on disk anyway.
6. By default, this flag is set to 120 seconds. Setting it to zero disables it, i.e. you can have a permanently busy priority queue.
7. I am acutely aware / wary of the fact that this separation-of-the-queues-plus-config-param is purely solving a problem that is (for me at least) purely a theoretically-possible issue. I've never seen it, but I don't think it's premature optimisation to tackle it in this change.

## Thing 2: Handling the fact that SLAs stop the standard queue from ever being empty.
1. As noted ... somewhere ... above, a hacky fix was to simply stop adding dags to the queue on SLA callback. But this wasn't perfect for people who don't happen to use airflow the way I do.
2. So, the first half of Thing 2 was to make the standard queue FIFO. This is effectively upholding the change from the previous MR; SLA callbacks simply get treated the same as dags loaded from disk, and have to wait their turn.
3. The next bit was to then add a `set` which tracked which dag files in the queue were still outstanding from the last refresh from disk. Once the queue was refreshed from disk, we'd work through every file eventually (because FIFO), and at that point the set would be empty, even if the queue wasn't (because SLAs).
4. Once the set was empty, we'd refresh the queue with dags found on disk.
5. Some of those dags from disk _might already be in the queue due to SLAs_ but we don't care; this is a good thing as it means loading dags from disk doesn't push existing SLAs to the back of the line. 

# Notes

This doesn't materially change _how_ SLAs work; they are generated and consumed the same as before. It just means that we reliably consume the alerts once generated without breaking the rest of the system. In my experience at least, adding SLAs would simply cause the system to stop processing dag updates (as per #15596).

In particular, I don't address issues with SLA timestamps (as raised by #22532), nor do I deal with other problems (e.g. now SLAs fire reliably, I have noticed that they fire during catch-up, and that the same alert can fire multiple times).

This is not because I think the current approach is perfect (like everyone else on the internet, I Have Thoughts on how it could be improved, given infinite time) but rather it is a sad-but-true fact that I don't have the time to take on a big project. So I am going to continue with my current approach of tinkering round the edges; and if the remaining issues are minor enough to live with, I'm just going to live with them (at least for now).

# Summary: I'm from the UK; politely waiting in line is what we do best.

- Separation of manager's `_file_path_queue` to `_std_file_path_queue` (for SLAs and dags loaded from disk) and `priority_file_path_queue` (for DAG call backs)
- New `max_file_process_interval` config to ensure files are read from disk every so often, even if the priority queue is always busy
- Dags from disk will be added to std queue even if the std queue is still busy with SLAs.
- Some extra stats/metrics to track queue depths / progress
- Tests updated to reflect changes

### CODE REVIEW

**Comment by argibbs** [comment]
@potiuk you may also be interested in this one.

File: airflow/dag_processing/manager.py [hunk 1]
```diff
@@ -378,8 +378,17 @@ def __init__(
...
+        self._log = logging.getLogger('airflow.processor_manager')
+
+        # known files; this will be updated every `dag_dir_list_interval` and stuff added/removed accordingly
         self._file_paths: List[str] = []
-        self._file_path_queue: List[str] = []
+
+        # we maintain 2 queues: stuff requiring rapid response due to scheduler updates, and stuff that
+        # should be serviced once the priority stuff has all been worked through, e.g. periodic dir scans
+        # additionally there's a set to track which files on disk still haven't been refreshed yet
+        self._priority_file_path_queue: Deque[str] = deque()
```
Comment (argibbs) [inline airflow/dag_processing/manager.py]: I changed these from `list`s to `deque`s because teeeeechnically it's more efficient.

Also, I like seeing people try to work out how to pronounce 'deque'.

File: airflow/dag_processing/manager.py [hunk 2]
```diff
@@ -378,8 +378,17 @@ def __init__(
         async_mode: bool = True,
     ):
         super().__init__()
+        self._log = logging.getLogger('airflow.processor_manager')
```
Comment (argibbs) [inline airflow/dag_processing/manager.py]: I moved this up from further down `__init__` because I added a log message which came before it in the old location.

It seems ... weird ... that this doesn't just use the logging mixin. Is this an artefact from Ye Olden Times?

File: airflow/dag_processing/manager.py [hunk 3]
```diff
@@ -539,7 +559,7 @@ def _run_parsing_loop(self):
             poll_time = None
 
         self._refresh_dag_dir()
-        self.prepare_file_path_queue()
+        self.populate_std_file_queue_from_dir()
```
Comment (argibbs) [inline airflow/dag_processing/manager.py]: Same method, different name. I thought this was more descriptive.

File: airflow/dag_processing/manager.py [hunk 4]
```diff
@@ -921,8 +944,18 @@ def set_file_paths(self, new_file_paths):
...
         self._file_paths = new_file_paths
-        self._file_path_queue = [x for x in self._file_path_queue if x in new_file_paths]
+
+        # clean up the queues; remove anything queued which no longer in the list, including callbacks
+        self._priority_file_path_queue = deque(
+            x for x in self._priority_file_path_queue if x in new_file_paths
+        )
+        self._std_file_path_queue = deque(x for x in self._std_file_path_queue if x in new_file_paths)
+        callback_paths_to_del = list(x for x in self._callback_to_execute.keys() if x not in new_file_paths)
+        for path_to_del in callback_paths_to_del:
```
Comment (argibbs) [inline airflow/dag_processing/manager.py]: This more thorough clean-up is new. Having the old callbacks lying around was a form of memory leak, but in practice not a serious one.

File: tests/dag_processing/test_manager.py [hunk 5]
```diff
@@ -299,9 +299,9 @@ def test_file_paths_in_queue_sorted_alphabetically(
         )
 
         manager.set_file_paths(dag_files)
-        assert manager._file_path_queue == []
-        manager.prepare_file_path_queue()
-        assert manager._file_path_queue == ['file_1.py', 'file_2.py', 'file_3.py', 'file_4.py']
+        assert list(manager._std_file_path_queue) == []
```
Comment (argibbs) [inline tests/dag_processing/test_manager.py]: Have to convert all the `deque`s to lists. This is clunky.

File: airflow/config_templates/config.yml [hunk 6]
```diff
@@ -1913,6 +1913,21 @@
...
+        Number of seconds since the previous parse time after which a DAG file will be parsed
+        again. This is only useful if your system contains a lot of dags / slas which update
+        exceedingly frequently, as this may prevent the system from scanning the file system
+        for changes to your dag files. This operates on a best efforts basis; if you have
+        many dags, and it takes 10 minutes to scan and parse them all, this will not make
+        it parse any more often.
+        Setting this to <= 0 disables the behaviour, in case it's important to you that those
+        frequently updating dags / slas always take priority at the cost of delaying updates
+        from disk
+      version_added: 2.4.0
```
Comment (argibbs) [inline airflow/config_templates/config.yml]: I mean, I'm assuming that this PR will be "subject to discussion" to put it mildly and has no chance of making 2.3.4. If it sails through, I'll happily correct this.

File: newsfragments/25489.bugfix.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+DAGProcessorManager queue made fully FIFO, split into two (priority and std), and dags guaranteed to be read from disk periodically (fixes/improves SLA alert reliability)
```
Comment (argibbs) [inline newsfragments/25489.bugfix.rst]: I suspect "bugfix" undersells the change. As per PR description, will happily change the classification to whatever I'm told is the right one to use for a change like this.

File: airflow/dag_processing/manager.py [hunk 8]
```diff
@@ -600,11 +620,14 @@ def _run_parsing_loop(self):
...
 
-            # Generate more file paths to process if we processed all the files
-            # already.
-            if not self._file_path_queue:
+            # Generate more file paths to process if we processed all the files already. Note for this
+            # to clear down, we must have first either:
+            # 1. cleared the priority queue then cleared this set while draining the std queue, or
+            # 2. been unable to clear the priority queue, hit max_file_process_interval, and drained the set
+            #    while clearing overdue files
+            if not self._outstanding_std_file_paths:
```
Comment (argibbs) [inline airflow/dag_processing/manager.py]: This is a key change as mentioned in the PR description: we refresh files from disk once the _`set`_ is empty. The queue might still have SLA callbacks in it, but that shouldn't stop us refreshing from disk.

File: airflow/dag_processing/manager.py [hunk 9]
```diff
@@ -993,19 +1027,56 @@ def _create_process(file_path, pickle_dags, dag_ids, callback_requests):
...
+
     def start_new_processes(self):
         """Start more processors if we have enough slots and files to process"""
-        while self._parallelism - len(self._processors) > 0 and self._file_path_queue:
-            file_path = self._file_path_queue.pop(0)
-            # Stop creating duplicate processor i.e. processor with the same filepath
+        # check for files stuck on the std queue for too long; these need to be processed first, as they
+        # suggest that we found files on disk a long time ago and still haven't processed it
+        now = timezone.utcnow()
+        overdue_std_files_list = [p for p in self._file_paths if self.is_overdue(p, now)]
```
Comment (argibbs) [inline airflow/dag_processing/manager.py:1044]: Note we're checking all known files (`self._file_paths`) rather than the std queue. This is because if the priority queue is spinning, then the std queue won't empty, and won't reload from disk, and there might be files on disk not in the std queue.

**Comment by potiuk** [comment]
Thanks for so thorough description and your willingness to improve SLA (I really, really appreciate it as this is one of the things that we had on our backburner for quite a while).

However, I think @argibbs this is the kind of change and desciption is such big and potential of breaking things is such big that it requires a devlist discussion IMHO to drag attention of people who should be dragged rather than dicussing it only in PR.  I know you wrote ("probably not make it in 2.3.4"), but I also think "probably not make it in 2.4.0 either" is a better assesment.

I strongly suggest to make a small digest of the description you did - with most important parts of "why" and "how" extracted (less is more) and send it to the devlist https://airflow.apache.org/community/, with link to this PR and inviting people to express their thoughts.

I think this is not going to be merged before 2.4 branch is cut. Too big risks, it needs to be likely very thoroughly tested and justified that we know what we are doing here) so there is stiil a lot of time to discuss it. 

I also think some people who are now looking at finalizing [AIP-48](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-48+Data+Dependency+Management+and+Data+Driven+Scheduling) so that it is ready for prime time for 2.4 (and who should be part of the discussiosn), have no time and energy to look at it. And we definiitely want to avoid of adding too many potentially high-impact changes in one release. We've already added too much in 2.3.0 and I think we learned a lesson there.

So regardless of what you propose it might take a while

I strongly encourage you to describe the case/problem/solution in concise way and send it to the devlist.

**Comment by potiuk** [comment]
But I do feel that direction you took is sound. We need some prioritisation of callbacks. It just need deeper look and discussions that fits a PR opened maybe 2 weeks before substantial release which is already undergoing quite substantial testing.

**Comment by argibbs** [comment]
Sounds good to me. I'm running with this locally, so I don't really mind about timelines. I'm just trying to give back (and also push upstream so I don't have to keep patching each time there's a release :smile: )

Will do as you suggested re: the digest etc.

**Comment by potiuk** [comment]
> Sounds good to me. I'm running with this locally, so I don't really mind about timelines. I'm just trying to give back (and also push upstream so I don't have to keep patching each time there's a release smile )

Cool! This is the RIGHT approach! Much appreciated!

**Comment by Jorricks** [comment]
@argibbs these changes look great! Very well described as well.

>  am acutely aware / wary of the fact that this separation-of-the-queues-plus-config-param is purely solving a problem that is (for me at least) purely a theoretically-possible issue. I've never seen it, but I don't think it's premature optimisation to tackle it in this change.

We have had the case where because of a crucial singel point of failure, hardly any DAG was able to continue (as we use a lot of ExternalTaskSensors). This meant that all of our schedulers were down and handling SLA fires for a long time so I am very happy to see this optimisation 👍 

### Note: I don't think this should be included in this MR but given you worked on this, you probably have a good view on this.
I was thinking. Part of what we can speed up is to instead of fire each SLA for the same DAG individually, group SLAs per DAG. Thereby, we add a dag with an SLA to the queue instead of an actual SLA. This prevents us from parsing a DAG file 10 times, for 10 different SLAs of the same DAG. I guess you could optimise this by including a dummy operator at the end that requires all the important tasks to succeed, but that's not super user friendly.

**Comment by argibbs** [comment]
@Jorricks thank you for the kind words.

I agree, SLAs could/should be made more efficient. I found this change easier/simpler to make & test, possibly because it's a problem I've tackled many times in my "real" job. Also, as you note, I think this change defends against a range of possible problems, rather than just SLAs.

I may try to improve SLA firing in the future, but I will be upfront about my motivations; I'm not looking to be "The SLA guy", as I simply don't have the spare time. If I end up with a load of spare time, that will doubtless change...

**Comment by potiuk** [comment]
I think it is a bit risky for 2.4 - so this one might stay open for a while until we branch off for 2.4 (@argibbs - it's not forgotten I can assure you).

**Comment by potiuk** [comment]
Some rebase is needed after the __future__annotation change, but I think we are pretty close to the "focused" 2.4.0 release effort. I think this is a good time to rebase/fix and maybe re-raise a devlist discussion about it.

**Comment by argibbs** [comment]
Hiya, yes, I hadn't forgotten either. :smile: 

Have just found some time to refresh this, and am sorting the rebase now. Will re-raise the dev-list email once that's done (and I'll spell your name right this time too!)

**Comment by argibbs** [comment]
Hiya, yes, I hadn't forgotten either. :smile: 

Have just found some time to refresh this, and am sorting the rebase now. Will re-raise the dev-list email once that's done (and I'll spell your name right this time too!)

**Comment by argibbs** [comment]
I have no idea why, but after rebasing, github seems convinced there's no changed in the branch, and has closed and won't let me reopen it ... I've created a new PR (from the same branch - see #27317 - it's the exact same change)
